backend/module/chaos/serverexperimentation/server_experimentation_integration_test.go

```diff
@@ -0,0 +1,233 @@
+// to run assertions against.
+func TestEnvoyRTDSFaults(t *testing.T) {
+	xdsConfig := &xdsconfigv1.Config{
+		RtdsLayerName: "rtds",
+	}
+	seConfig := &serverexperimentationconfigv1.Config{
+		IngressFaultRuntimePrefix: "fault.http",
+	}
+
+	ts, err := newServerExperimentationTestModuleServer(seConfig, xdsConfig)
+	assert.NoError(t, err)
+	defer ts.Stop()
+
+	e, err := envoytest.NewEnvoyHandle()
+	assert.NoError(t, err)
+
+	e.EnsureControlPlaneConnectivity(envoytest.RuntimeStatPrefix)
+	assert.NoError(t, err)
+
+	code, err := e.MakeSimpleCall()
+	assert.NoError(t, err)
+	assert.Equal(t, 503, code)
+
+	createTestExperiment(t, 400, ts.Storer)
+
+	err = awaitExpectedReturnValueForSimpleCall(t, e, awaitReturnValueParams{
+		timeout:        4 * time.Second,
+		expectedStatus: 400,
+	})
+	assert.NoError(t, err, "did not see faults enabled")
+
+	// We know that faults have been applied, now try to kill the server and ensure that we eventually reset the faults.
+	// This verifies that we're properly setting TTLs and that Envoy will honor this.
+	ts.Stop()
+
+	err = awaitExpectedReturnValueForSimpleCall(t, e, awaitReturnValueParams{
+		timeout:        10 * time.Second,
+		expectedStatus: 503,
+	})
+	assert.NoError(t, err, "did not see faults reverted")
+}
+
+func TestEnvoyECDSFaults(t *testing.T) {
+	xdsConfig := &xdsconfigv1.Config{
+		RtdsLayerName:        "rtds",
+		CacheRefreshInterval: durationpb.New(time.Second),
+		EcdsAllowList:        &xdsconfigv1.Config_ECDSAllowList{EnabledClusters: []string{"test-cluster"}},
+	}
+	seConfig := &serverexperimentationconfigv1.Config{}
+
+	ts, err := newServerExperimentationTestModuleServer(seConfig, xdsConfig)
+	assert.NoError(t, err)
+	defer ts.Stop()
+
+	e, err := envoytest.NewEnvoyHandle()
+	assert.NoError(t, err)
+
+	err = e.EnsureControlPlaneConnectivity(envoytest.EcdsStatPrefix)
+	assert.NoError(t, err)
+
+	code, err := e.MakeSimpleCall()
+	assert.NoError(t, err)
+	assert.Equal(t, 503, code)
+
+	experiment := createTestExperiment(t, 404, ts.Storer)
+
+	err = awaitExpectedReturnValueForSimpleCall(t, e, awaitReturnValueParams{
+		timeout:        4 * time.Second,
+		expectedStatus: 404,
+	})
+	assert.NoError(t, err, "did not see faults enabled")
+
+	// TODO(kathan24): Test TTL by stopping the server instead of canceling the experiment. Currently, TTL is not not supported for ECDS in the upstream Envoy
+	ts.Storer.CancelExperimentRun(context.Background(), experiment.Run.Id, "")
+
+	err = awaitExpectedReturnValueForSimpleCall(t, e, awaitReturnValueParams{
+		timeout:        10 * time.Second,
+		expectedStatus: 503,
+	})
+	assert.NoError(t, err, "did not see faults reverted")
+}
+
+// Helper class for initializing a test RTDS server.
+type TestModuleServer struct {
+	*xdstest.TestServer
+
+	Storer *experimentstoremock.SimpleStorer
+}
+
+func newServerExperimentationTestModuleServer(seConfig *serverexperimentationconfigv1.Config, xdsConfig *xdsconfigv1.Config) (*TestModuleServer, error) {
```

Comment:
this method is a light modification of `NewModuleTest` method that existed previously (and still exists in `xdstest` package)